File: algorithms/medium/ones_and_zeros.ex
defmodule Solution do
  @spec find_max_form(strs :: [String.t()], m :: integer, n :: integer) :: integer
  def find_max_form(strs, m, n) do
    for s <- strs, reduce: %{} do
      dp ->
        {ones, zeros} = ones_and_zeros(s)

        for i <- m..zeros//-1, j <- n..ones//-1, reduce: dp do
          dp ->
            Map.put(
              dp,
              {i, j},
              max(Map.get(dp, {i, j}, 0), 1 + Map.get(dp, {i - zeros, j - ones}, 0))
            )
        end
    end
    |> Map.get({m, n}, 0)
  end

  def ones_and_zeros(s) do
    s
    |> String.graphemes()
    |> Enum.reduce({0, 0}, fn c, {ones, zeros} ->
      if c == "1" do
        {ones + 1, zeros}
      else
        {ones, zeros + 1}
      end
    end)
  end
end

File: algorithms/medium/ones_and_zeros.rs
fn ones_and_zeros (s: &str) -> (usize, usize) {
    s.as_bytes().iter().fold((0, 0), |(z, o), c| {
        if *c == b'0' {
            (z + 1, o)
        } else {
            (z, o + 1)
        }
    })
}

impl Solution {
    pub fn find_max_form(strs: Vec<String>, m: i32, n: i32) -> i32 {
        let m = m as usize;
        let n = n as usize;
        let mut dp = vec![vec![0; n + 1]; m + 1];
        for s in &strs {
            let (zs, os) = ones_and_zeros(s);
            for i in (zs..=m).rev() {
                for j in (os..=n).rev() {
                    dp[i][j] = dp[i][j].max(1 + dp[i-zs][j-os]);
                }
            }
        }
        dp[m][n]
    }
}
